### ISSUE TITLE
Moshi support fails when serializing collections

### ISSUE THREAD

**Issue by karottenreibe** [http4k/http4k#78]
Example JUnit 5 Test that fails:

```kotlin
class MoshiTest {

    @Test
    fun `will throw an exception`() {
        val moshi = com.squareup.moshi.Moshi.Builder().build()
        // this works and outputs ["foo"] as expected
        println(moshi.adapter(List::class.java).failOnUnknown().toJson(listOf("foo")))

        // this fails
        println(org.http4k.format.Moshi.asJsonString(listOf("foo")))
    }

}
```

The error message is `java.lang.IllegalArgumentException: Platform class java.util.Collections$SingletonList annotated [] requires explicit JsonAdapter to be registered`

The same exception is thrown when trying to use a Lens for a collection class

The same if you pass multiple list elements, except it will be ArrayList that it complains about. The root cause is that http4k's Moshi wrapper always passes an object's actual class to Moshi instead of the interface class (List in this case). Moshi, however, will not serialize implementations, it deliberately only serializes against the interface in order to avoid tieing us to one specific implementation on one specific JVM version.

This probably affects other collections like maps etc as well.

__Expected__: http4k's Moshi adapter allows me to pass the desired interface, e.g. List::class, explicitly to properly serialize the collection.

__Context__: http4k 3.6.0, Moshi 1.5.0

**Comment by daviddenton** [comment]
On further inspection, this seems to be the same issue that we have when deserialising raw lists using the Jackson auto-conversion. Lists do seem to behave correctly when nested inside other objects.

We found that the way around this with Jackson was to use arrays instead of raw lists - this also worked with the lens mechanism and works with Moshi:
```Body.auto<Array<Whatever>>().toLens()```

Without using the array workaround, attempting to deserialise a data class like this:
```data class ArbObject(val string: String, val child: ArbObject?, val numbers: List<Int>, val bool: Boolean)````
.. using a standard Moshi adapter fromJson() call results in an `ArrayList<LinkedTreeHashMap>` being created.

I've also added the missing fromJsonString method to Moshi which allows you to pass the class to the serialisation method should the above
